atef/widgets/happi.py

```diff
@@ -0,0 +1,369 @@
+    """
+    Happi item (device) search widget.
+
+    Parameters
+    ----------
+    parent : QWidget, optional
+        The parent widget.
+
+    client : happi.Client, optional
+        Happi client instance.  May be supplied at initialization time or
+        later.
+    """
+    filename: ClassVar[str] = 'happi_search_widget.ui'
+    happi_items_selected: ClassVar[QtCore.Signal] = QtCore.Signal(list)
+
+    _client: Optional[happi.client.Client]
+    combo_by_category: QtWidgets.QComboBox
+    device_selection_group: QtWidgets.QGroupBox
+    layout_by_name: QtWidgets.QHBoxLayout
+    radio_by_category: QtWidgets.QRadioButton
+    radio_by_name: QtWidgets.QRadioButton
+    list_or_tree_frame: QtWidgets.QFrame
+    button_refresh: QtWidgets.QPushButton
+    happi_tree_view: HappiDeviceTreeView
+    happi_list_view: HappiDeviceListView
+    _tree_updated: bool
+    _tree_current_category: str
+    _search_thread: Optional[ThreadWorker]
+
+    def __init__(
+        self,
+        parent: Optional[QWidget] = None,
+        *,
+        client: Optional[happi.Client] = None,
+    ):
+        super().__init__(parent=parent)
+        self._client = None
+        self._tree_current_category = "beamline"
+        self._search_thread = None
+        self._tree_has_data = False
+        self._setup_ui()
+        # Set the client at the end, as this may trigger an update:
+        self.client = client
+
+    def _setup_ui(self):
+        self.happi_tree_view.setVisible(False)
+        self._setup_tree_view()
+
+        self.button_refresh.clicked.connect(self.refresh_happi)
+        self.list_or_tree_frame.layout().insertWidget(0, self.happi_list_view)
+
+        self.radio_by_name.clicked.connect(self._select_device_widget)
+        self.radio_by_category.clicked.connect(self._select_device_widget)
+        self.combo_by_category.currentTextChanged.connect(self._category_changed)
+        self.button_refresh.clicked.emit()
+
+        def list_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            self.happi_items_selected.emit([idx.data() for idx in selected.indexes()])
+
+        self.happi_list_view.selectionModel().selectionChanged.connect(
+            list_selection_changed
+        )
+
+    def _setup_tree_view(self):
+        """Set up the happi_tree_view if not already configured."""
+        view = self.happi_tree_view
+        view.groups = [
+            self.combo_by_category.itemText(idx)
+            for idx in range(self.combo_by_category.count())
+        ]
+        self.list_or_tree_frame.layout().insertWidget(0, view)
+
+        def tree_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            items = [
+                idx.data() for idx in selected.indexes()
+                if idx.parent().data() is not None  # skip top-level items
+            ]
+            self.happi_items_selected.emit(items)
```

Comment:
Is this where we would get the happi info out of this dialog and into a field somewhere in the config gui? What exactly do we get here and why is it a list?

Reply:
A list is fine, I just didn't see where the multi-select was currently being used.

Somewhat related, the type information is really opaque with the `QtCore.Signal(list)`. If it's strings that's great, if it's containers that'd also be great, it isn't obvious with my partial understanding.

Reply:
Agreed on the opaqueness. 

I initially had the `Signal` with `List[str]`, but PyQt doesn't want annotations, it wants objects. 🤷‍♂️ 
I wonder if either pyqt or pyside will eventually have better support in this regard.

Reply:
Ugh, qt probably wants you to use QStringList

Reply:
Ahh, true - I totally forgot about that...

Reply:
Wait, QStringList doesn't exist in PyQt5 any more! It went away with `QString`, thankfully. I guess I'll stick with `list` and a comment about the annotation.

Reply:
Oh thank goodness I never want to see it again